libclc/ptx-nvidiacl/libspirv/images/image.cl

```diff
@@ -2588,3 +2599,349 @@ _CLC_DEFINE_MIPMAP_BINDLESS_READS_BUILTIN(half4, 3, Dv4_DF16_, v4f16, Dv4_f, flo
 #undef COORD_PARAMS_3D
 
 #undef _CLC_DEFINE_MIPMAP_BINDLESS_READS_BUILTIN
+
+// ------- Image Arrays / Layered Images -------
+
+// //  --- THUNKS: Surface Array Reads ---
+// int
+int __nvvm_suld_1d_array_i32_clamp_s(long, int, int) __asm(
+    "llvm.nvvm.suld.1d.array.i32.clamp");
+int __nvvm_suld_2d_array_i32_clamp_s(long, int, int, int) __asm(
+    "llvm.nvvm.suld.2d.array.i32.clamp");
+int2 __nvvm_suld_1d_array_v2i32_clamp_s(long, int, int) __asm(
+    "__clc_llvm_nvvm_suld_1d_array_v2i32_clamp");
+int2 __nvvm_suld_2d_array_v2i32_clamp_s(long, int, int, int) __asm(
+    "__clc_llvm_nvvm_suld_2d_array_v2i32_clamp");
+int4 __nvvm_suld_1d_array_v4i32_clamp_s(long, int, int) __asm(
+    "__clc_llvm_nvvm_suld_1d_array_v4i32_clamp");
+int4 __nvvm_suld_2d_array_v4i32_clamp_s(long, int, int, int) __asm(
+    "__clc_llvm_nvvm_suld_2d_array_v4i32_clamp");
+
+// short
+short __nvvm_suld_1d_array_i16_clamp_s(long, int, int) __asm(
+    "llvm.nvvm.suld.1d.array.i16.clamp");
+short __nvvm_suld_2d_array_i16_clamp_s(long, int, int, int) __asm(
+    "llvm.nvvm.suld.2d.array.i16.clamp");
+short2 __nvvm_suld_1d_array_v2i16_clamp_s(long, int, int) __asm(
+    "__clc_llvm_nvvm_suld_1d_array_v2i16_clamp");
+short2 __nvvm_suld_2d_array_v2i16_clamp_s(long, int, int, int) __asm(
+    "__clc_llvm_nvvm_suld_2d_array_v2i16_clamp");
+short4 __nvvm_suld_1d_array_v4i16_clamp_s(long, int, int) __asm(
+    "__clc_llvm_nvvm_suld_1d_array_v4i16_clamp");
+short4 __nvvm_suld_2d_array_v4i16_clamp_s(long, int, int, int) __asm(
+    "__clc_llvm_nvvm_suld_2d_array_v4i16_clamp");
+
+// char helper -- i8 intrinsic returns i16, requires helper
+short __nvvm_suld_1d_array_i8_clamp_s_helper(long, int, int) __asm(
+    "llvm.nvvm.suld.1d.array.i8.clamp");
+short __nvvm_suld_2d_array_i8_clamp_s_helper(long, int, int, int) __asm(
+    "llvm.nvvm.suld.2d.array.i8.clamp");
+short2 __nvvm_suld_1d_array_v2i8_clamp_s_helper(long, int, int) __asm(
```

Comment:
With 4 cases of `__nvvm_suld_[ARRAY_D]_array_[TYPE]_clamp_s_helper` I'd be tempted to wrap it in a macro, such is libclc style anyway 😄